### PR TITLE
Board valid_placement?

### PR DESCRIPTION
**Pull request by whitneykidd**
Created valid_placement? with passing tests. In order to do this, I created methods that check for valid horizontal placement and valid vertical placement, and a method that checks that all passed in coordinates are empty.

### CODE REVIEW

File: lib/board.rb [hunk 1]
```diff
@@ -29,4 +29,50 @@ def valid_coordinate?(coordinate)
...
+      coordinate2_number = coordinate2[1].to_i
+
+      coordinate1_letter == coordinate2_letter &&
+      coordinate1_number == (coordinate2_number - 1)
+    end
+  end
+
+  def place_vertical?(coordinates)
+    coordinates.each_cons(2).all? do |coordinate1, coordinate2|
+      coordinate1_letter = coordinate1[0].ord #A of A1
```
Comment (whitneykidd) [inline lib/board.rb:64]: I left these comments in for reference because this was tricky for me to wrap my head around.

File: lib/board.rb [hunk 2]
```diff
@@ -29,4 +29,50 @@ def valid_coordinate?(coordinate)
       true
     end
   end
+
+  def valid_placement?(ship, coordinates)
```
Comment (whitneykidd) [inline lib/board.rb:33]: This method is definitely a candidate for refactoring if/when we get to that.

File: lib/board.rb [hunk 3]
```diff
@@ -29,4 +29,50 @@ def valid_coordinate?(coordinate)
...
+      if ship.length == coordinates.length && coordinates.all? { |coordinate| valid_coordinate? (coordinate) }
```
Comment (melatran) [inline lib/board.rb:34]: Notes to myself to remember what it's doing:
This first checks the length where the number of coordinates should equal the length of the ship.
.all? method returns true when all elements in array are true

File: lib/board.rb [hunk 4]
```diff
@@ -29,4 +29,50 @@ def valid_coordinate?(coordinate)
...
+        if place_ship?(coordinates) == true && is_occupied?(coordinates) == false
+          true
+        else
+          false
+        end
+      else
```
Comment (melatran) [inline lib/board.rb:40]: I don't think we need this second else statement.

File: lib/board.rb [hunk 5]
```diff
@@ -29,4 +29,50 @@ def valid_coordinate?(coordinate)
...
+        false
+      end
+    end
+
+  def is_occupied?(coordinates)
+    coordinates.any? { |coordinate| !@cells[coordinate].empty? }
```
Comment (melatran) [inline lib/board.rb:46]: notes to myself:
.any? returns true if at least one element in the block is true
! bang and does the opposite "Are any of the cells not empty?"

File: lib/board.rb [hunk 6]
```diff
@@ -29,4 +29,50 @@ def valid_coordinate?(coordinate)
...
+  end
+
+  def place_horizontal?(coordinates)
+    coordinates.each_cons(2).all? do |coordinate1, coordinate2|
```
Comment (melatran) [inline lib/board.rb:50]: note to myself:
.each_cons iterates over consecutive elements

File: lib/board.rb [hunk 7]
```diff
@@ -29,4 +29,50 @@ def valid_coordinate?(coordinate)
...
+
+      coordinate2_letter = coordinate2[0].ord
+      coordinate2_number = coordinate2[1].to_i
+
+      coordinate1_letter == coordinate2_letter &&
+      coordinate1_number == (coordinate2_number - 1)
+    end
+  end
+
+  def place_vertical?(coordinates)
```
Comment (melatran) [inline lib/board.rb:62]: I like the separate place horizontal and vertical methods. it's organized well and it makes sense.

File: test/board_test.rb [hunk 8]
```diff
@@ -24,4 +26,25 @@ def test_is_validates_coordinates
...
+
+  def test_valid_placement
+    assert_equal false, @board.valid_placement?(@cruiser, ["A1", "A2", "A4"])
+    assert_equal false, @board.valid_placement?(@submarine, ["A1", "C1"])
+    assert_equal false, @board.valid_placement?(@cruiser, ["A3", "A2", "A1"])
+    assert_equal false, @board.valid_placement?(@submarine, ["C1", "B1"])
+    assert_equal false, @board.valid_placement?(@cruiser, ["A1", "B2", "C3"])
+  end
+
+  def test_can_place_horizontal
```
Comment (melatran) [inline test/board_test.rb:43]: I also like these tests for horizontal and vertical.

File: test/board_test.rb [hunk 9]
```diff
@@ -24,4 +26,25 @@ def test_is_validates_coordinates
...
     assert_equal false, @board.valid_coordinate?("A22")
   end
+
+  def test_returns_valid_placement_length
+    assert_equal false, @board.valid_placement?(@cruiser, ["A1", "A2"])
+    assert_equal false, @board.valid_placement?(@submarine, ["A2", "A3", "A4"])
+  end
+
+  def test_valid_placement
+    assert_equal false, @board.valid_placement?(@cruiser, ["A1", "A2", "A4"])
```
Comment (melatran) [inline test/board_test.rb:36]: I like these tests that shows what should happen when the coordinates aren't consecutive.

File: lib/board.rb (hunk 4 above)
Comment (whitneykidd) [reply lib/board.rb:40]: If it's taken out, test_returns_valid_placement_length returns nil instead of false.

Comment (melatran) [inline lib/board.rb:40]: note to myself: this else is with the first "if"

File: lib/board.rb [hunk 10]
```diff
@@ -29,4 +29,50 @@ def valid_coordinate?(coordinate)
       true
     end
   end
+
+  def valid_placement?(ship, coordinates)
+      if ship.length == coordinates.length && coordinates.all? { |coordinate| valid_coordinate? (coordinate) }
+        if place_ship?(coordinates) == true && is_occupied?(coordinates) == false
+          true
+        else
```
Comment (melatran) [inline lib/board.rb:37]: notes to myself:
this else is with the first 'if'

**Pull request merged**